core/lib/multivm/src/versions/vm_latest/tracers/default_tracers.rs

```diff
@@ -44,27 +44,125 @@ pub(crate) struct DefaultExecutionTracer<S: WriteStorage, H: HistoryMode> {
+    pub(crate) fn new(
+        computational_gas_limit: u32,
+        execution_mode: VmExecutionMode,
+        dispatcher: TracerDispatcher<S, H>,
+        storage: StoragePtr<S>,
+        refund_tracer: Option<RefundsTracer<S>>,
+        pubdata_tracer: Option<PubdataTracer<S>>,
+    ) -> Self {
+        Self {
+            tx_has_been_processed: false,
+            execution_mode,
+            gas_spent_on_bytecodes_and_long_messages: 0,
+            computational_gas_used: 0,
+            tx_validation_gas_limit: computational_gas_limit,
+            in_account_validation: false,
+            final_batch_info_requested: false,
+            result_tracer: ResultTracer::new(execution_mode),
+            refund_tracer,
+            dispatcher,
+            pubdata_tracer,
+            ret_from_the_bootloader: None,
+            storage,
+            _phantom: Default::default(),
+        }
+    }
+
+    pub(crate) fn tx_has_been_processed(&self) -> bool {
+        self.tx_has_been_processed
+    }
+
+    pub(crate) fn validation_run_out_of_gas(&self) -> bool {
+        self.computational_gas_used > self.tx_validation_gas_limit
+    }
+
+    pub(crate) fn gas_spent_on_pubdata(&self, vm_local_state: &VmLocalState) -> u32 {
+        self.gas_spent_on_bytecodes_and_long_messages + vm_local_state.spent_pubdata_counter
+    }
+
+    fn set_fictive_l2_block(
+        &mut self,
+        state: &mut ZkSyncVmState<S, H>,
+        bootloader_state: &mut BootloaderState,
+    ) {
+        let current_timestamp = Timestamp(state.local_state.timestamp);
+        let txs_index = bootloader_state.free_tx_index();
+        let l2_block = bootloader_state.insert_fictive_l2_block();
+        let mut memory = vec![];
+        apply_l2_block(&mut memory, l2_block, txs_index);
+        state
+            .memory
+            .populate_page(BOOTLOADER_HEAP_PAGE as usize, memory, current_timestamp);
+        self.final_batch_info_requested = false;
+    }
+
+    fn should_stop_execution(&self) -> TracerExecutionStatus {
+        match self.execution_mode {
+            VmExecutionMode::OneTx if self.tx_has_been_processed() => {
+                return TracerExecutionStatus::Stop(TracerExecutionStopReason::Finish);
+            }
+            VmExecutionMode::Bootloader if self.ret_from_the_bootloader == Some(RetOpcode::Ok) => {
+                return TracerExecutionStatus::Stop(TracerExecutionStopReason::Finish);
+            }
+            _ => {}
+        };
+        if self.validation_run_out_of_gas() {
+            return TracerExecutionStatus::Stop(TracerExecutionStopReason::Abort(
+                Halt::ValidationOutOfGas,
+            ));
+        }
+        TracerExecutionStatus::Continue
+    }
+}
+
 impl<S: WriteStorage, H: HistoryMode> Debug for DefaultExecutionTracer<S, H> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("DefaultExecutionTracer").finish()
     }
 }
 
+macro_rules! dispatch_tracers {
```

Comment:
Actually, that exactly what we want to avoid.. 

The initial idea was, that we have all tracers including the service one such as result together in one vector. But it caused serious performance degradation ~5% for each tracer.. That's why we've started  to move service tracers into default tracer.

Reply:
Is the performance regression caused by dynamic dispatch? Didn't think it could have that significant impact. Dynamic dispatch could be replaced with a static dispatch via an enum, but it'd look no better than the current solution. Alternatively, one could try emulating `impl for<T: VmTracer<S, H>> FnMut(&mut T)`, but that also looks quite awkward.

Reply:
but it made =( 

So I don't think the current macros is bad :) and it seem to meet our needs.